File: rn/apps/windows/src/native/windows/CoreIPCModule.cpp
#include "pch.h"
#include "CoreIPCModule.h"

namespace LiteLLMMenu {

void CoreIPCModule::Initialize(winrt::Microsoft::ReactNative::ReactContext const& context) noexcept {
  try {
    context_ = context;
    auto dispatcher = context_.JSDispatcher();
    auto emitter = CoreEvent;
    CoreIPCBridge::Shared().SetEventHandler([dispatcher, emitter](std::string const& event) {
      dispatcher.Post([emitter, event] { emitter(event); });
    });
  } catch (...) {
  }
}

void CoreIPCModule::Send(
    std::string request,
    winrt::Microsoft::ReactNative::ReactPromise<std::string> promise) noexcept {
  try {
    auto dispatcher = context_.JSDispatcher();
    std::thread([request = std::move(request), promise = std::move(promise), dispatcher]() mutable {
      try {
        std::string response = CoreIPCBridge::Shared().Send(request);
        dispatcher.Post([promise = std::move(promise), response = std::move(response)] { promise.Resolve(response); });
      } catch (...) {
        dispatcher.Post([promise = std::move(promise)] { promise.Reject("The local Core is unavailable."); });
      }
    }).detach();
  } catch (...) {
    promise.Reject("The local Core is unavailable.");
  }
}

void CoreIPCModule::Shutdown() noexcept {
  try {
    CoreIPCBridge::Shared().Stop();
  } catch (...) {
  }
}

void CoreIPCModule::AddListener(std::string const&) noexcept {}
void CoreIPCModule::RemoveListeners(double) noexcept {}

}  // namespace LiteLLMMenu

File: rn/apps/macos/src/native/macos/CoreIPCModule.swift
import Foundation

#if canImport(React)
import React

@objc(LiteLLMCore)
final class LiteLLMCoreModule: RCTEventEmitter {
    private let core = CoreIPCBridge.shared
    private var observing = false

    override init() {
        super.init()
        core.setEventHandler { [weak self] event in
            DispatchQueue.main.async {
                guard let self, self.observing else { return }
                self.sendEvent(withName: "coreEvent", body: event)
            }
        }
    }

    override static func requiresMainQueueSetup() -> Bool { true }
    override func supportedEvents() -> [String]! { ["coreEvent"] }
    override func startObserving() { observing = true }
    override func stopObserving() { observing = false }

    @objc func send(_ request: String, resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        core.send(request) { result in
            DispatchQueue.main.async {
                switch result {
                case .success(let response): resolve(response)
                case .failure: reject("core_unavailable", "The local Core is unavailable.", nil)
                }
            }
        }
    }

    @objc func shutdown() { core.stop() }
}
#endif
